zerver/actions/users.py

```diff
@@ -376,54 +376,63 @@ def send_stream_events_for_role_update(
 def do_change_user_role(
     user_profile: UserProfile, value: int, *, acting_user: Optional[UserProfile]
 ) -> None:
-    old_value = user_profile.role
-    old_system_group = get_system_user_group_for_user(user_profile)
+    # We re-fetch the user for here, inside the transaction, so we can
+    # lock it and know its current role for certain, without races.
+    locked_user_profile = UserProfile.objects.select_for_update().get(id=user_profile.id)
```

Comment:
hmm, could we just keep it called `user_profile` here? Or perhaps better, change the parameter we accept to be called `unlocked_user_profile`? 

As written, I feel like someone is inevitably going to accidentally access the original `user_profile` object in future refactoring out of habit.

Reply:
I originally wrote it that way, for precisely that reason.  But there is a bunch of code which relies on the `user_profile` that it passes in having an updated `.role`.  There is no way to `user_profile.select_for_update().refresh_from_db()` (ref https://code.djangoproject.com/ticket/28344).

It doesn't seem necessarily cleaner, but we could also:
```python
passed_in_user_profile = user_profile
user_profile = UserProfile.objects.select_for_update().get(id=user_profile.id)
...
user_profile.role = ...
...
passed_in_user_profile.refresh_from_db()
```

Reply:
Pushed a revised version which should be cleaner.